### ISSUE TITLE
Terraform 0.13 Data Source Read Causes Unexpected Plan Difference

### ISSUE THREAD

**Issue by bflad** [hashicorp/terraform-plugin-testing#239]
### terraform-plugin-testing version

```
v1.6.0
```

### Relevant provider source code

```go
func TestAccDataSourceCertificate_UpgradeFromVersion3_4_0(t *testing.T) {
	r.Test(t, r.TestCase{
		Steps: []r.TestStep{
			{
				ExternalProviders: providerVersion340(),
				Config: `
					data "tls_certificate" "test" {
						content = file("fixtures/certificate.pem")
					}
				`,
				Check: r.ComposeAggregateTestCheckFunc(
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.#", "1"),

					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.signature_algorithm", "SHA256-RSA"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.public_key_algorithm", "RSA"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.serial_number", "266244246501122064554217434340898012243"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.is_ca", "false"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.version", "3"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.issuer", "CN=Root CA,O=Test Org,L=Here"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.subject", "CN=Child Cert,O=Child Co.,L=Everywhere"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_before", "2019-11-08T09:01:36Z"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_after", "2019-11-08T19:01:36Z"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.sha1_fingerprint", "61b65624427d75b61169100836904e44364df817"),
					tu.TestCheckPEMFormat("data.tls_certificate.test", "certificates.0.cert_pem", PreambleCertificate.String()),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.cert_pem", strings.TrimSpace(fixtures.TestTlsDataSourceCertFromContent)+"\n"),
				),
			},
			{
				ProtoV5ProviderFactories: protoV5ProviderFactories(),
				Config: `
					data "tls_certificate" "test" {
						content = file("fixtures/certificate.pem")
					}
				`,
				PlanOnly: true,
			},
			{
				ProtoV5ProviderFactories: protoV5ProviderFactories(),
				Config: `
					data "tls_certificate" "test" {
						content = file("fixtures/certificate.pem")
					}
				`,
				Check: r.ComposeAggregateTestCheckFunc(
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.#", "1"),

					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.signature_algorithm", "SHA256-RSA"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.public_key_algorithm", "RSA"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.serial_number", "266244246501122064554217434340898012243"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.is_ca", "false"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.version", "3"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.issuer", "CN=Root CA,O=Test Org,L=Here"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.subject", "CN=Child Cert,O=Child Co.,L=Everywhere"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_before", "2019-11-08T09:01:36Z"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_after", "2019-11-08T19:01:36Z"),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.sha1_fingerprint", "61b65624427d75b61169100836904e44364df817"),
					tu.TestCheckPEMFormat("data.tls_certificate.test", "certificates.0.cert_pem", PreambleCertificate.String()),
					r.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.cert_pem", strings.TrimSpace(fixtures.TestTlsDataSourceCertFromContent)+"\n"),
				),
			},
		},
	})
}
```

### Terraform Configuration Files

Reproducible testing a data source by itself:

```tf
data "tls_certificate" "test" {
  content = file("fixtures/certificate.pem")
}
```

### Expected Behavior

Either guidance on how to resolve the issue in the v1.6.0 release notes or the testing to pass on Terraform 0.13 as it did previously.

### Actual Behavior

Worked on terraform-plugin-testing v1.5.1, then upon trying to upgrade to v1.6.0, new test failure:

```
=== RUN   TestAccDataSourceCertificate_UpgradeFromVersion3_4_0
    data_source_certificate_test.go:51: Step 2/3 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
         <= read (data resources)
        
        Terraform will perform the following actions:
        
          # data.tls_certificate.test will be read during apply
          # (config refers to values not yet known)
         <= data "tls_certificate" "test"  {
                certificates = [
                    {
                        cert_pem             = <<~EOT
                            -----BEGIN CERTIFICATE-----
                            MIIDUzCCAjugAwIBAgIRAMhMxtTgTXHTmo6ZU7OafFMwDQYJKoZIhvcNAQELBQAw
                            NDENMAsGA1UEBxMESGVyZTERMA8GA1UEChMIVGVzdCBPcmcxEDAOBgNVBAMTB1Jv
                            b3QgQ0EwHhcNMTkxMTA4MDkwMTM2WhcNMTkxMTA4MTkwMTM2WjA+MRMwEQYDVQQH
                            EwpFdmVyeXdoZXJlMRIwEAYDVQQKEwlDaGlsZCBDby4xEzARBgNVBAMTCkNoaWxk
                            IENlcnQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDQTeCu4[66](https://github.com/hashicorp/terraform-provider-tls/actions/runs/7090729003/job/19298389836?pr=449#step:5:67)xxnGr
                            CCrl823J4gGnp9AYb0laTP3uB4orXblTFq45ehDnEJXNykT+7acT8IrAjQlVQdl0
                            gLjNM6XjGkFQ7xRw5xi041vRrOtUzC1KxVqrcfT4WrKj6zM/MuK3hznc4NvvwdAx
                            Mb3Sk46yQ1PrMslsidDvhTAqXkVi3lD1bV/bpnDo3NRCldVpedE1wlR+6thXZN/Y
                            MggNuDdv6LDadVGlXgKw5KkEIgenGOzpX1o+GKGo5UWu1xoTHikVwEC1iVuCZax+
                            9FnHQO/q7SyF4Lb9d0j6vzrIAjzauGbiAsJya1GhYMF7INxzpSolzk0UYjT5Dxcq
                            d3VX1prxAgMBAAGjVjBUMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEF
                            BQcDATAMBgNVHRMBAf8EAjAAMB8GA1UdIwQYMBaAFIBRoM9+w7/obXaqAmaCPyVf
                            ldxEMA0GCSqGSIb3DQEBCwUAA4IBAQCuXJkT+qD3STmyDlsJOQRLBKaECH+/0mw4
                            mn3oMikNfneybjhao+fpwTgFup3KIrdIgbuciHfSTZzWT6mDs9bUdZZLccU6cVRh
                            WiX0I1eppjQyOT7PuXDsOsBUMf+et5WuGYrtKsib07q2rHPtTq72iftANtWbznfq
                            DsM3TQL4LuEE9V2lU2L2f3kXKrkYzLJj7R4sGck5Fo/E8eeIFm1Z5FCPcia82N+C
                            xDsNFvV3r8TsRH60IxFekKddI+ivepa97SvC4r+[69](https://github.com/hashicorp/terraform-provider-tls/actions/runs/7090729003/job/19298389836?pr=449#step:5:70)MPyxULHNwDtSL+8T4q01LEP
                            VKT7dWjBK3K0xxH0SPCtlqRbGalWz4adNNHazN/x7ebK+WB9ReSM
                            -----END CERTIFICATE-----
                        EOT
                        is_ca                = false
                        issuer               = "CN=Root CA,O=Test Org,L=Here"
                        not_after            = "2019-11-08T19:01:36Z"
                        not_before           = "2019-11-08T09:01:36Z"
                        public_key_algorithm = "RSA"
                        serial_number        = "266244246501122064554217434340898012243"
                        sha1_fingerprint     = "61b65624427d75b61169100836904e44364df817"
                        signature_algorithm  = "SHA256-RSA"
                        subject              = "CN=Child Cert,O=Child Co.,L=Everywhere"
                        version              = 3
                    },
                ]
                content      = <<~EOT
                    -----BEGIN CERTIFICATE-----
                    MIIDUzCCAjugAwIBAgIRAMhMxtTgTXHTmo6ZU7OafFMwDQYJKoZIhvcNAQELBQAw
                    NDENMAsGA1UEBxMESGVyZTERMA8GA1UEChMIVGVzdCBPcmcxEDAOBgNVBAMTB1Jv
                    b3QgQ0EwHhcNMTkxMTA4MDkwMTM2WhcNMTkxMTA4MTkwMTM2WjA+MRMwEQYDVQQH
                    EwpFdmVyeXdoZXJlMRIwEAYDVQQKEwlDaGlsZCBDby4xEzARBgNVBAMTCkNoaWxk
                    IENlcnQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDQTeCu466xxnGr
                    CCrl823J4gGnp9AYb0laTP3uB4orXblTFq45ehDnEJXNykT+7acT8IrAjQlVQdl0
                    gLjNM6XjGkFQ7xRw5xi041vRrOtUzC1KxVqrcfT4WrKj6zM/MuK3hznc4NvvwdAx
                    Mb3Sk46yQ1PrMslsidDvhTAqXkVi3lD1bV/bpnDo3NRCldVpedE1wlR+6thXZN/Y
                    MggNuDdv6LDadVGlXgKw5KkEIgenGOzpX1o+GKGo5UWu1xoTHikVwEC1iVuCZax+
                    9FnHQO/q7SyF4Lb9d0j6vzrIAjzauGbiAsJya1GhYMF7INxzpSolzk0UYjT5Dxcq
                    d3VX1prxAgMBAAGjVjBUMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEF
                    BQcDATAMBgNVHRMBAf8EAjAAMB8GA1UdIwQYMBaAFIBRoM9+w7/obXaqAmaCPyVf
                    ldxEMA0GCSqGSIb3DQEBCwUAA4IBAQCuXJkT+qD3STmyDlsJOQRLBKaECH+/0mw4
                    mn3oMikNfneybjhao+fpwTgFup3KIrdIgbuciHfSTZzWT6mDs9bUdZZLccU6cVRh
                    WiX0I1eppjQyOT7PuXDsOsBUMf+et5WuGYrtKsib07q2rHPtTq72iftANtWbznfq
                    DsM3TQL4LuEE9V2lU2L2f3kXKrkYzLJj7R4sGck5Fo/E8eeIFm1Z5FCPcia82N+C
                    xDsNFvV3r8TsRH60IxFekKddI+ivepa97SvC4r+69MPyxULHNwDtSL+8T4q01LEP
                    VKT7dWjBK3K0xxH0SPCtlqRbGalWz4adNNHazN/x7ebK+WB9ReSM
                    -----END CERTIFICATE-----
                EOT
              ~ id           = "3412221cbf0c80eb3f50040db8b20b98336bb57f" -> "b80a5a1bf12b39c0c44ff25e1c1b[71](https://github.com/hashicorp/terraform-provider-tls/actions/runs/7090729003/job/19298389836?pr=449#step:5:72)e5d1c4990a"
                verify_chain = true
            }
        
        Plan: 0 to add, 0 to change, 0 to destroy.
--- FAIL: TestAccDataSourceCertificate_UpgradeFromVersion3_4_0 (2.72s)
```

### Steps to Reproduce

1. In terraform-provider-tls: `TF_ACC=1 TF_ACC_TERRAFORM_PATH=/path/to/terraform0.13.x go test -count=1 -run='TestAccDataSourceCertificate_UpgradeFromVersion3_4_0' -v ./internal/provider`

### References

- https://github.com/hashicorp/terraform-provider-tls/pull/449

**Comment by bflad** [comment]
Additional context: Terraform 0.13 had a few quirks with its data source handling behavior and apparently this extended to how plans were generated as well. These issues were mainly resolved in Terraform 0.14.

Some workaround or potentially recommended options:

* Use `TerraformVersionChecks` with `tfversion.SkipBetween(tfversion.Version_0_13_0, tfversion.Version_0_14_0)` to skip Terraform 0.13 testing
* Remove `PlanOnly: true`
* Add `ExpectNonEmptyPlan: true`

To adjust the testing logic, it might be a matter of skipping the non-refresh plan testing if Terraform 0.13 is detected.

**Comment by bflad** [comment]
While a very targeted fix for this situation could be created, this is likely a very niche case since running a refresh plan is more common practitioner behavior and this error is due to a non-refresh plan on Terraform 0.13. Most providers are not expected to be testing on Terraform 0.13 these days and this situation this arises is very specific (e.g. on non-refresh plan with PlanOnly and config has data sources with slightly changing values). So rather than potentially introduce any risk and additional complexity with a targeted fix, the recommendation for now will be to add `TestCase.TerraformVersionChecks` with `tfversion.SkipBetween(tfversion.Version_0_13_0, tfversion.Version_0_14_0)`.

If more developers run into this issue and it is a larger scale problem, we can revisit this decision. 👍
